File: package/main/src/DataStructure/priorityQueue.ts
/**
 * A priority queue implementation using a binary heap.
 * Higher priority values are dequeued first.
 * 
 * ## Features
 * - **enqueue(value, priority)**: Add element with priority
 * - **enqueueBack(value)**: Add element to the back with lowest priority
 * - **dequeue()**: Remove and return highest priority element
 * - **peek()**: View highest priority element without removing
 * - **peekPriority()**: View highest priority value
 * - **size**: Get number of elements
 * - **isEmpty**: Check if queue is empty
 * - **clear()**: Remove all elements
 * - **toArray()**: Get all elements as array
 * - **toArrayWithPriorities()**: Get all elements with priorities
 * 
 * ## Time Complexity
 * - enqueue: O(log n)
 * - dequeue: O(log n)
 * - peek: O(1)
 * - peekPriority: O(1)
 * 
 * @example
 * ```typescript
 * const queue = new PriorityQueue<string>();
 * queue.enqueue("low", 1);
 * queue.enqueue("high", 3);
 * queue.enqueue("medium", 2);
 * 
 * console.log(queue.dequeue()); // "high"
 * console.log(queue.dequeue()); // "medium"
 * console.log(queue.dequeue()); // "low"
 * ```
 * 
 * @example
 * ```typescript
 * // Initialize with elements
 * const queue = new PriorityQueue([
 *   { value: "task1", priority: 1 },
 *   { value: "task2", priority: 3 },
 *   { value: "task3", priority: 2 }
 * ]);
 * 
 * console.log(queue.peek()); // "task2"
 * console.log(queue.peekPriority()); // 3
 * ```
 * 
 * @template T - The type of elements stored in the queue
 */
export class PriorityQueue<T> {
  private heap: Array<{ value: T; priority: number }> = [];

  /**
   * Creates a new PriorityQueue instance.
   * @param initialElements - Optional array of initial elements with priorities
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * // or
   * const queue = new PriorityQueue([
   *   { value: "item1", priority: 10 },
   *   { value: "item2", priority: 5 }
   * ]);
   * ```
   */
  constructor(initialElements?: Array<{ value: T; priority: number }>) {
    if (initialElements) {
      this.heap = [...initialElements];
      this.buildHeap();
    }
  }

  /**
   * Returns the number of elements in the queue.
   * @returns The number of elements in the queue
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("item", 1);
   * console.log(queue.size); // 1
   * ```
   */
  get size(): number {
    return this.heap.length;
  }

  /**
   * Checks if the queue is empty.
   * @returns True if the queue is empty, false otherwise
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * console.log(queue.isEmpty); // true
   * queue.enqueue("item", 1);
   * console.log(queue.isEmpty); // false
   * ```
   */
  get isEmpty(): boolean {
    return this.heap.length === 0;
  }

  /**
   * Adds an element to the queue with a specified priority.
   * Higher priority values are dequeued first.
   * @param value - The value to add
   * @param priority - The priority value (higher values have higher priority)
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("low", 1);
   * queue.enqueue("high", 10);
   * queue.enqueue("medium", 5);
   * console.log(queue.dequeue()); // "high"
   * ```
   */
  enqueue(value: T, priority: number): void {
    this.heap.push({ value, priority });
    this.heapifyUp(this.heap.length - 1);
  }

  /**
   * Adds an element to the end of the queue with lowest priority.
   * This element will be dequeued last (FIFO for equal lowest priority).
   * @param value - The value to add
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("high", 10);
   * queue.enqueue("medium", 5);
   * queue.enqueueBack("back1");
   * queue.enqueueBack("back2");
   * console.log(queue.dequeue()); // "high"
   * console.log(queue.dequeue()); // "medium"
   * console.log(queue.dequeue()); // "back1"
   * console.log(queue.dequeue()); // "back2"
   * ```
   */
  enqueueBack(value: T): void {
    const lowestPriority = this.getLowestPriority();
    const newPriority = lowestPriority - 1;
    this.heap.push({ value, priority: newPriority });
    this.heapifyUp(this.heap.length - 1);
  }

  /**
   * Removes and returns the element with the highest priority.
   * @returns The element with highest priority, or undefined if queue is empty
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("low", 1);
   * queue.enqueue("high", 10);
   * console.log(queue.dequeue()); // "high"
   * console.log(queue.dequeue()); // "low"
   * console.log(queue.dequeue()); // undefined
   * ```
   */
  dequeue(): T | undefined {
    if (this.heap.length === 0) {
      return;
    }

    if (this.heap.length === 1) {
      return this.heap.pop()?.value;
    }

    const result = this.heap[0].value;
    // biome-ignore lint/style/noNonNullAssertion: pop() cannot return undefined when heap.length > 1
    this.heap[0] = this.heap.pop()!;
    this.heapifyDown(0);
    return result;
  }

  /**
   * Returns the element with the highest priority without removing it.
   * @returns The element with highest priority, or undefined if queue is empty
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("low", 1);
   * queue.enqueue("high", 10);
   * console.log(queue.peek()); // "high"
   * console.log(queue.size); // 2 (element not removed)
   * ```
   */
  peek(): T | undefined {
    return this.heap[0]?.value;
  }

  /**
   * Returns the priority of the element with the highest priority.
   * @returns The highest priority value, or undefined if queue is empty
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("low", 1);
   * queue.enqueue("high", 10);
   * console.log(queue.peekPriority()); // 10
   * ```
   */
  peekPriority(): number | undefined {
    return this.heap[0]?.priority;
  }

  /**
   * Removes all elements from the queue.
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("item1", 1);
   * queue.enqueue("item2", 2);
   * console.log(queue.size); // 2
   * queue.clear();
   * console.log(queue.size); // 0
   * console.log(queue.isEmpty); // true
   * ```
   */
  clear(): void {
    this.heap = [];
  }

  /**
   * Returns an array of all elements in the queue (without removing them).
   * The order is not guaranteed to be sorted by priority.
   * @returns Array of all elements in the queue
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("low", 1);
   * queue.enqueue("high", 10);
   * queue.enqueue("medium", 5);
   * console.log(queue.toArray()); // ["high", "medium", "low"] (order may vary)
   * ```
   */
  toArray(): T[] {
    return this.heap.map((item) => item.value);
  }

  /**
   * Returns an array of all elements with their priorities.
   * The order is not guaranteed to be sorted by priority.
   * @returns Array of all elements with their priorities
   * @example
   * ```typescript
   * const queue = new PriorityQueue<string>();
   * queue.enqueue("low", 1);
   * queue.enqueue("high", 10);
   * console.log(queue.toArrayWithPriorities()); 
   * // [{ value: "high", priority: 10 }, { value: "low", priority: 1 }] (order may vary)
   * ```
   */
  toArrayWithPriorities(): Array<{ value: T; priority: number }> {
    return [...this.heap];
  }

  /**
   * Gets the lowest priority value in the queue.
   * @returns The lowest priority value, or 0 if queue is empty
   */
  private getLowestPriority(): number {
    if (this.heap.length === 0) {
      return 0;
    }
    return Math.min(...this.heap.map(item => item.priority));
  }

  /**
   * Builds a max heap from the current heap array.
   */
  private buildHeap(): void {
    for (
      let index = Math.floor(this.heap.length / 2) - 1;
      index >= 0;
      index--
    ) {
      this.heapifyDown(index);
    }
  }

  /**
   * Moves an element up the heap to maintain heap property.
   * @param index - The index of the element to move up
   */
  private heapifyUp(index: number): void {
    let currentIndex = index;
    while (currentIndex > 0) {
      const parentIndex = Math.floor((currentIndex - 1) / 2);
      if (this.heap[currentIndex].priority <= this.heap[parentIndex].priority) {
        break;
      }
      this.swap(currentIndex, parentIndex);
      currentIndex = parentIndex;
    }
  }

  /**
   * Moves an element down the heap to maintain heap property.
   * @param index - The index of the element to move down
   */
  private heapifyDown(index: number): void {
    let currentIndex = index;
    while (true) {
      const leftChild = 2 * currentIndex + 1;
      const rightChild = 2 * currentIndex + 2;
      let largest = currentIndex;

      if (
        leftChild < this.heap.length &&
        this.heap[leftChild].priority > this.heap[largest].priority
      ) {
        largest = leftChild;
      }

      if (
        rightChild < this.heap.length &&
        this.heap[rightChild].priority > this.heap[largest].priority
      ) {
        largest = rightChild;
      }

      if (largest === currentIndex) {
        break;
      }

      this.swap(currentIndex, largest);
      currentIndex = largest;
    }
  }

  /**
   * Swaps two elements in the heap.
   * @param i - First index
   * @param j - Second index
   */
  private swap(index: number, index_: number): void {
    [this.heap[index], this.heap[index_]] = [
      this.heap[index_],
      this.heap[index],
    ];
  }
}

File: package/umt_go/src/datastructure/priority_queue.go
package datastructure

// PriorityQueue is a generic priority queue backed by a binary heap.
// The ordering is determined by a comparison function provided at construction.
//
// The compare function should return:
//   - a negative value if a has higher priority than b (a should be dequeued first)
//   - zero if a and b have equal priority
//   - a positive value if b has higher priority than a (b should be dequeued first)
//
// This follows the convention where compare(a, b) < 0 means a comes first,
// similar to sort.Slice's less function but returning an int.
type PriorityQueue[T any] struct {
	heap    []T
	compare func(a, b T) int
}

// NewPriorityQueue creates a new PriorityQueue with the given comparison function.
//
// The compare function determines element ordering:
//   - compare(a, b) < 0: a has higher priority (dequeued first)
//   - compare(a, b) == 0: equal priority
//   - compare(a, b) > 0: b has higher priority (dequeued first)
//
// Example (max-heap of ints):
//
//	pq := NewPriorityQueue(func(a, b int) int { return b - a })
//
// Example (min-heap of ints):
//
//	pq := NewPriorityQueue(func(a, b int) int { return a - b })
func NewPriorityQueue[T any](compare func(a, b T) int) *PriorityQueue[T] {
	return &PriorityQueue[T]{
		heap:    make([]T, 0),
		compare: compare,
	}
}

// Push adds an item to the priority queue. The item is placed according to
// its priority as determined by the comparison function. O(log n).
func (pq *PriorityQueue[T]) Push(item T) {
	pq.heap = append(pq.heap, item)
	pq.heapifyUp(len(pq.heap) - 1)
}

// Pop removes and returns the highest-priority item from the queue.
// The second return value is false if the queue is empty.
// O(log n).
func (pq *PriorityQueue[T]) Pop() (T, bool) {
	if len(pq.heap) == 0 {
		var zero T
		return zero, false
	}

	if len(pq.heap) == 1 {
		item := pq.heap[0]
		pq.heap = pq.heap[:0]
		return item, true
	}

	item := pq.heap[0]
	last := len(pq.heap) - 1
	pq.heap[0] = pq.heap[last]
	pq.heap = pq.heap[:last]
	pq.heapifyDown(0)
	return item, true
}

// Peek returns the highest-priority item without removing it.
// The second return value is false if the queue is empty. O(1).
func (pq *PriorityQueue[T]) Peek() (T, bool) {
	if len(pq.heap) == 0 {
		var zero T
		return zero, false
	}
	return pq.heap[0], true
}

// Size returns the number of items in the queue.
func (pq *PriorityQueue[T]) Size() int {
	return len(pq.heap)
}

// IsEmpty returns true if the queue contains no items.
func (pq *PriorityQueue[T]) IsEmpty() bool {
	return len(pq.heap) == 0
}

// heapifyUp restores the heap property by moving the element at index up
// towards the root.
func (pq *PriorityQueue[T]) heapifyUp(index int) {
	for index > 0 {
		parent := (index - 1) / 2
		if pq.compare(pq.heap[index], pq.heap[parent]) >= 0 {
			break
		}
		pq.heap[index], pq.heap[parent] = pq.heap[parent], pq.heap[index]
		index = parent
	}
}

// heapifyDown restores the heap property by moving the element at index down
// towards the leaves.
func (pq *PriorityQueue[T]) heapifyDown(index int) {
	n := len(pq.heap)
	for {
		left := 2*index + 1
		right := 2*index + 2
		smallest := index

		if left < n && pq.compare(pq.heap[left], pq.heap[smallest]) < 0 {
			smallest = left
		}
		if right < n && pq.compare(pq.heap[right], pq.heap[smallest]) < 0 {
			smallest = right
		}

		if smallest == index {
			break
		}

		pq.heap[index], pq.heap[smallest] = pq.heap[smallest], pq.heap[index]
		index = smallest
	}
}
